contracts/src/v0.8/upkeeps/ERC20BalanceMonitor.sol

```diff
@@ -0,0 +1,289 @@
+    uint96 topUpAmount;
+    uint56 lastTopUpTimestamp;
+  }
+
+  address private s_keeperRegistryAddress;
+  uint256 private s_minWaitPeriodSeconds;
+  address[] private s_watchList;
+  mapping(address => Target) internal s_targets;
+
+  /**
+   * @param erc20TokenAddress the ERC20 token address
+   * @param keeperRegistryAddress the address of the keeper registry contract
+   * @param minWaitPeriodSeconds the minimum wait period for addresses between funding
+   */
+  constructor(
+    address erc20TokenAddress,
+    address keeperRegistryAddress,
+    uint256 minWaitPeriodSeconds
+  ) ConfirmedOwner(msg.sender) {
+    setERC20TokenAddress(erc20TokenAddress);
+    setKeeperRegistryAddress(keeperRegistryAddress);
+    setMinWaitPeriodSeconds(minWaitPeriodSeconds);
+  }
+
+  /**
+   * @notice Sets the list of subscriptions to watch and their funding parameters.
+   * @param addresses the list of subscription ids to watch
+   * @param minBalances the minimum balances for each subscription
+   * @param topUpAmounts the amount to top up each subscription
+   */
+  function setWatchList(
+    address[] calldata addresses,
+    uint96[] calldata minBalances,
+    uint96[] calldata topUpAmounts
+  ) external onlyOwner {
+    if (addresses.length != minBalances.length || addresses.length != topUpAmounts.length) {
+      revert InvalidWatchList();
+    }
+    address[] memory oldWatchList = s_watchList;
+    for (uint256 idx = 0; idx < oldWatchList.length; idx++) {
+      s_targets[oldWatchList[idx]].isActive = false;
+    }
+    for (uint256 idx = 0; idx < addresses.length; idx++) {
+      if (s_targets[addresses[idx]].isActive) {
+        revert DuplicateAddress(addresses[idx]);
+      }
+      if (addresses[idx] == address(0)) {
+        revert InvalidWatchList();
+      }
+      if (topUpAmounts[idx] == 0) {
+        revert InvalidWatchList();
+      }
+      s_targets[addresses[idx]] = Target({
+        isActive: true,
+        minBalance: minBalances[idx],
+        topUpAmount: topUpAmounts[idx],
+        lastTopUpTimestamp: 0
+      });
+    }
+    s_watchList = addresses;
+    emit WatchlistUpdated(oldWatchList, addresses);
+  }
+
+  /**
+   * @notice Gets a list of subscriptions that are underfunded.
+   * @return list of subscriptions that are underfunded
+   */
+  function getUnderfundedAddresses() public view returns (address[] memory) {
+    address[] memory watchList = s_watchList;
+    address[] memory needsFunding = new address[](watchList.length);
+    uint256 count = 0;
+    uint256 minWaitPeriod = s_minWaitPeriodSeconds;
+    uint256 contractBalance = ERC20Token.balanceOf(address(this));
+    Target memory target;
+    for (uint256 idx = 0; idx < watchList.length; idx++) {
+      target = s_targets[watchList[idx]];
+      uint256 targetTokenBalance = ERC20Token.balanceOf(watchList[idx]);
+      if (
+        target.lastTopUpTimestamp + minWaitPeriod <= block.timestamp &&
+        contractBalance >= target.topUpAmount &&
```

Comment:
might be useful to revert here instead of returning false as you can use it to differentiate between
- no address needing topUp
- addresses needing topUp but contract out of funds

Reply:
I think the optimal behavior for this contract is to have the keeper fund as many addresses as possible, even if it starts running out of gas. While this change would potentially add more versatility, it would reduce the speed of funding for our current use-case.

Reply:
yeah makes sense, I was just thinking of how you would get to know that this contract requires more funds. There would be other ways to do this too